tests/lldp/test_lldp.py - test failed due to unexpected peer MAC address in LLDP data

**Description**

tests/lldp/test_lldp.py - test failed due to unexpected peer MAC address in LLDP data. Issue happen after merge pull request: https://github.com/Azure/sonic-mgmt/pull/3331/
It looks like duthost.facts['router_mac'].replace(':', '') took router mac, but in LLDP packet we have mac address for MGMT(eth0) interface which is different than router mac.

**Steps to reproduce the issue:**
1. Run tests from file tests/lldp/test_lldp.py
2. Check result

**Describe the results you received:**
``` 
duthosts = <tests.common.devices.duthosts.DutHosts object at 0x7f0c0b34ff50> 

enum_rand_one_per_hwsku_frontend_hostname = 'r-tigon-11' 

localhost = <tests.common.devices.local.Localhost object at 0x7f0c0a43a6d0> 

eos = {'bgp_gr_timer': 700, 'ceos_image_mount_dir': '/data/ceos', 'snmp_location': 'str', 'snmp_rocommunity': 'public'} 

collect_techsupport_all_duts = None 

loganalyzer = {'r-tigon-11': <tests.common.plugins.loganalyzer.loganalyzer.LogAnalyzer instance at 0x7f0c08b21b90>} 

enum_frontend_asic_index = None 

 

    def test_lldp_neighbor(duthosts, enum_rand_one_per_hwsku_frontend_hostname, localhost, eos, 

                           collect_techsupport_all_duts, loganalyzer, enum_frontend_asic_index): 

        """ verify LLDP information on neighbors """ 

        duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname] 

 

        if loganalyzer: 

            loganalyzer[enum_rand_one_per_hwsku_frontend_hostname].ignore_regex.extend([ 

                ".*ERR syncd#syncd: :- check_fdb_event_notification_data.*", 

                ".*ERR syncd#syncd: :- process_on_fdb_event: invalid OIDs in fdb \ 

                    notifications, NOT translating and NOT storing in ASIC DB.*", 

                ".*ERR syncd#syncd: :- process_on_fdb_event: FDB notification was \ 

                    not sent since it contain invalid OIDs, bug.*", 

            ]) 

 

        res = duthost.shell("docker exec -i lldp lldpcli show chassis | grep \"SysDescr:\" | sed -e 's/^\\s*SysDescr:\\s*//g'") 

        dut_system_description = res['stdout'] 

        lldpctl_facts = duthost.lldpctl_facts(asic_instance_id=enum_frontend_asic_index, skip_interface_pattern_list=["eth0", "Ethernet-BP", "Ethernet-IB"])['ansible_facts'] 

        config_facts = duthost.asic_instance(enum_frontend_asic_index).config_facts(host=duthost.hostname, source="running")['ansible_facts'] 

 

        nei_meta = config_facts.get('DEVICE_NEIGHBOR_METADATA', {}) 

 

        for k, v in lldpctl_facts['lldpctl'].items(): 

            try: 

                hostip = v['chassis']['mgmt-ip'] 

            except: 

                logger.info("Neighbor device {} does not sent management IP via lldp".format(v['chassis']['name'])) 

                hostip = nei_meta[v['chassis']['name']]['mgmt_addr'] 

 

            nei_lldp_facts = localhost.lldp_facts(host=hostip, version='v2c', community=eos['snmp_rocommunity'])['ansible_facts'] 

            neighbor_interface = v['port']['ifname'] 

            # Verify the published DUT system name field is correct 

            assert nei_lldp_facts['ansible_lldp_facts'][neighbor_interface]['neighbor_sys_name'] == duthost.hostname 

            # Verify the published DUT chassis id field is not empty 

            assert nei_lldp_facts['ansible_lldp_facts'][neighbor_interface]['neighbor_chassis_id'] == \ 

>                  "0x%s" % (duthost.facts['router_mac'].replace(':', '')) 

E           AssertionError 

 

collect_techsupport_all_duts = None 

config_facts = {'ACL_TABLE': {'DATAACL': {'policy_desc': u'DATAACL', 'ports': [u'PortChannel0001', u'PortChannel0002', u'PortChannel0..._pool': {'mode': u'dynamic', 'type': u'ingress'}, 'ingress_lossy_pool': {'mode': u'dynamic', 'type': u'ingress'}}, ...} 

dut_system_description = u'SONiC Software Version: SONiC.202012.86-7424a137_Internal - HwSku: ACS-MSN4600C - Distribution: Debian 10.9 - Kernel: 4.19.0-12-2-amd64' 

duthost    = <MultiAsicSonicHost> r-tigon-11 

duthosts   = <tests.common.devices.duthosts.DutHosts object at 0x7f0c0b34ff50> 

enum_frontend_asic_index = None 

enum_rand_one_per_hwsku_frontend_hostname = 'r-tigon-11' 

eos        = {'bgp_gr_timer': 700, 'ceos_image_mount_dir': '/data/ceos', 'snmp_location': 'str', 'snmp_rocommunity': 'public'} 

hostip     = u'10.213.14.68' 

k          = 'Ethernet288' 

lldpctl_facts = {'lldpctl': {'Ethernet0': {'age': u'0 day, 00:47:13', 'chassis': {'Bridge': {'enabled': u'on'}, 'Router': {'enabled': ...}, 'port': {'aggregation': u'1000001', 'ifname': u'Ethernet1', 'mfs': u'9236', 'ttl': u'120'}, 'rid': u'4', ...}, ...}} 

localhost  = <tests.common.devices.local.Localhost object at 0x7f0c0a43a6d0> 

loganalyzer = {'r-tigon-11': <tests.common.plugins.loganalyzer.loganalyzer.LogAnalyzer instance at 0x7f0c08b21b90>} 

nei_lldp_facts = {'ansible_lldp_facts': {'Ethernet1': {'neighbor_chassis_id': u'0x0c42a1940a0c', 'neighbor_port_desc': u'ARISTA02T0:Eth...ONiC.202012.86-7424a137_Internal - HwSku: ACS-MSN4600C - Distribution: Debian 10.9 - Kernel: 4.19.0-12-2-amd64', ...}}} 

nei_meta   = {'ARISTA01T0': {'hwsku': u'Arista-VM', 'lo_addr': u'None', 'mgmt_addr': u'10.213.14.67', 'type': u'ToRRouter'}, 'ARIST...'}, 'ARISTA03T0': {'hwsku': u'Arista-VM', 'lo_addr': u'None', 'mgmt_addr': u'10.213.14.69', 'type': u'ToRRouter'}, ...} 

neighbor_interface = u'Ethernet1' 

res        = {'stderr_lines': [], u'cmd': u'docker exec -i lldp lldpcli show chassis | grep...00C - Distribution: Debian 10.9 - Kernel: 4.19.0-12-2-amd64'], 'failed': False} 

v          = {'age': u'0 day, 00:46:54', 'chassis': {'Bridge': {'enabled': u'on'}, 'Router': {'enabled': u'on'}, 'descr': u'Arista ...', ...}, 'port': {'aggregation': u'1000001', 'ifname': u'Ethernet1', 'mfs': u'9236', 'ttl': u'120'}, 'rid': u'25', ...} 

 

lldp/test_lldp.py:67: AssertionError 
```
 
**Describe the results you expected:**
Test passed

**Additional information you deem important:**

**Output of `show version`:**
 ```
             "SONiC Software Version: SONiC.202012.86-7424a137_Internal", 

            "Distribution: Debian 10.9", 

            "Kernel: 4.19.0-12-2-amd64", 

            "Build commit: 7424a137", 

            "Build date: Fri May 14 13:39:56 UTC 2021", 

            "Built by: sw-r2d2-bot@r-build-sonic-ci03", 

            "", 

            "Platform: x86_64-mlnx_msn4600c-r0", 

            "HwSKU: ACS-MSN4600C", 

            "ASIC: mellanox", 

            "ASIC Count: 1", 

            "Serial Number: MT2023X22082", 

            "Uptime: 22:55:17 up 47 min,  0 users,  load average: 0.78, 1.57, 1.76", 

            "", 

            "Docker images:", 

            "REPOSITORY                    TAG                           IMAGE ID            SIZE", 

            "docker-syncd-mlnx             202012.86-7424a137_Internal   af44b879d1b0        658MB", 

            "docker-syncd-mlnx             latest                        af44b879d1b0        658MB", 

            "docker-snmp                   202012.86-7424a137_Internal   d1ba66b12268        442MB", 

            "docker-snmp                   latest                        d1ba66b12268        442MB", 

            "docker-teamd                  202012.86-7424a137_Internal   bad9318400a1        411MB", 

            "docker-teamd                  latest                        bad9318400a1        411MB", 

            "docker-nat                    202012.86-7424a137_Internal   2cb661d0c0be        414MB", 

            "docker-nat                    latest                        2cb661d0c0be        414MB", 

            "docker-sonic-mgmt-framework   202012.86-7424a137_Internal   b9fa45c076f2        621MB", 

            "docker-sonic-mgmt-framework   latest                        b9fa45c076f2        621MB", 

            "docker-router-advertiser      202012.86-7424a137_Internal   7b48b7222bbd        400MB", 

            "docker-router-advertiser      latest                        7b48b7222bbd        400MB", 

            "docker-platform-monitor       202012.86-7424a137_Internal   ae1b602b5006        682MB", 

            "docker-platform-monitor       latest                        ae1b602b5006        682MB", 

            "docker-lldp                   202012.86-7424a137_Internal   e5959b8b6a55        440MB", 

            "docker-lldp                   latest                        e5959b8b6a55        440MB", 

            "docker-database               202012.86-7424a137_Internal   496359e87302        400MB", 

            "docker-database               latest                        496359e87302        400MB", 

            "docker-orchagent              202012.86-7424a137_Internal   26e85289bbba        429MB", 

            "docker-orchagent              latest                        26e85289bbba        429MB", 

            "docker-sonic-telemetry        202012.86-7424a137_Internal   2bf7cbd0c69f        490MB", 

            "docker-sonic-telemetry        latest                        2bf7cbd0c69f        490MB", 

            "docker-fpm-frr                202012.86-7424a137_Internal   fe0272e6e4b7        429MB", 

            "docker-fpm-frr                latest                        fe0272e6e4b7        429MB", 

            "docker-dhcp-relay             202012.86-7424a137_Internal   d71d87df3dbf        407MB", 

            "docker-dhcp-relay             latest                        d71d87df3dbf        407MB", 

            "docker-sflow                  202012.86-7424a137_Internal   6745af1b3b47        412MB", 

            "docker-sflow                  latest                        6745af1b3b47        412MB", 

            "docker-wjh                    202012.202012.0-409a727       d016b0e2c797        504MB", 

            "docker-wjh                    latest                        d016b0e2c797        504MB", 
```

--- a/tests/lldp/test_lldp.py
+++ b/tests/lldp/test_lldp.py
@@ -31,7 +31,7 @@ def test_lldp(duthosts, enum_rand_one_per_hwsku_frontend_hostname, localhost, co
 
 
 def test_lldp_neighbor(duthosts, enum_rand_one_per_hwsku_frontend_hostname, localhost, eos,
-                       collect_techsupport_all_duts, loganalyzer, enum_frontend_asic_index):
+                       collect_techsupport_all_duts, loganalyzer, enum_frontend_asic_index, tbinfo):
     """ verify LLDP information on neighbors """
     duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname]
 
@@ -48,7 +48,11 @@ def test_lldp_neighbor(duthosts, enum_rand_one_per_hwsku_frontend_hostname, loca
     dut_system_description = res['stdout']
     lldpctl_facts = duthost.lldpctl_facts(asic_instance_id=enum_frontend_asic_index, skip_interface_pattern_list=["eth0", "Ethernet-BP", "Ethernet-IB"])['ansible_facts']
     config_facts = duthost.asic_instance(enum_frontend_asic_index).config_facts(host=duthost.hostname, source="running")['ansible_facts']
- 
+    
+    # LLDPD use mac address of eth0 to generate chassis ID.
+    mgmt_alias = duthost.get_extended_minigraph_facts(tbinfo)["minigraph_mgmt_interface"]["alias"]
+    mgmt_mac = duthost.get_dut_iface_mac(mgmt_alias)
+
     nei_meta = config_facts.get('DEVICE_NEIGHBOR_METADATA', {})
 
     for k, v in lldpctl_facts['lldpctl'].items():
@@ -64,7 +68,7 @@ def test_lldp_neighbor(duthosts, enum_rand_one_per_hwsku_frontend_hostname, loca
         assert nei_lldp_facts['ansible_lldp_facts'][neighbor_interface]['neighbor_sys_name'] == duthost.hostname
         # Verify the published DUT chassis id field is not empty
         assert nei_lldp_facts['ansible_lldp_facts'][neighbor_interface]['neighbor_chassis_id'] == \
-               "0x%s" % (duthost.facts['router_mac'].replace(':', ''))
+               "0x%s" % (mgmt_mac.replace(':', ''))
         # Verify the published DUT system description field is correct
         assert nei_lldp_facts['ansible_lldp_facts'][neighbor_interface]['neighbor_sys_desc'] == dut_system_description
         # Verify the published DUT port id field is correct

--- a/tests/lldp/test_lldp.py
+++ b/tests/lldp/test_lldp.py
@@ -49,9 +49,14 @@ def test_lldp_neighbor(duthosts, enum_rand_one_per_hwsku_frontend_hostname, loca
     lldpctl_facts = duthost.lldpctl_facts(asic_instance_id=enum_frontend_asic_index, skip_interface_pattern_list=["eth0", "Ethernet-BP", "Ethernet-IB"])['ansible_facts']
     config_facts = duthost.asic_instance(enum_frontend_asic_index).config_facts(host=duthost.hostname, source="running")['ansible_facts']
     
-    # LLDPD use mac address of eth0 to generate chassis ID.
-    mgmt_alias = duthost.get_extended_minigraph_facts(tbinfo)["minigraph_mgmt_interface"]["alias"]
-    mgmt_mac = duthost.get_dut_iface_mac(mgmt_alias)
+    # We use the MAC of mgmt port to generate chassis ID as LLDPD dose. 
+    # To be compatible with PR #3331, we keep using router MAC on T2 devices
+    switch_mac = ""
+    if tbinfo["topo"]["type"] != "t2":
+        mgmt_alias = duthost.get_extended_minigraph_facts(tbinfo)["minigraph_mgmt_interface"]["alias"]
+        switch_mac = duthost.get_dut_iface_mac(mgmt_alias)
+    else:
+        switch_mac = duthost.facts['router_mac']
 
     nei_meta = config_facts.get('DEVICE_NEIGHBOR_METADATA', {})
 
@@ -68,7 +73,7 @@ def test_lldp_neighbor(duthosts, enum_rand_one_per_hwsku_frontend_hostname, loca
         assert nei_lldp_facts['ansible_lldp_facts'][neighbor_interface]['neighbor_sys_name'] == duthost.hostname
         # Verify the published DUT chassis id field is not empty
         assert nei_lldp_facts['ansible_lldp_facts'][neighbor_interface]['neighbor_chassis_id'] == \
-               "0x%s" % (mgmt_mac.replace(':', ''))
+               "0x%s" % (switch_mac.replace(':', ''))
         # Verify the published DUT system description field is correct
         assert nei_lldp_facts['ansible_lldp_facts'][neighbor_interface]['neighbor_sys_desc'] == dut_system_description
         # Verify the published DUT port id field is correct

